client/pkg/transport/listener_test.go

```diff
@@ -575,3 +588,182 @@ func TestSocktOptsEmpty(t *testing.T) {
+	return ca, caBytes, priv
+}
+
+func generateCerts(privKey *ecdsa.PrivateKey, rootCA *x509.Certificate, dir, suffix string, t *testing.T) {
+	priv, err := ecdsa.GenerateKey(elliptic.P521(), rand.Reader)
+	if err != nil {
+		t.Fatal(err)
+	}
+	serialNumber := newSerialNumber(t)
+	tmpl := x509.Certificate{
+		SerialNumber:          serialNumber,
+		Subject:               pkix.Name{Organization: []string{"etcd"}},
+		NotBefore:             time.Now(),
+		NotAfter:              time.Now().Add(365 * (24 * time.Hour)),
+		ExtKeyUsage:           []x509.ExtKeyUsage{x509.ExtKeyUsageServerAuth, x509.ExtKeyUsageClientAuth},
+		KeyUsage:              x509.KeyUsageKeyEncipherment | x509.KeyUsageDigitalSignature | x509.KeyUsageContentCommitment,
+		BasicConstraintsValid: true,
+		IPAddresses:           []net.IP{net.ParseIP("127.0.0.1")},
+	}
+	caBytes, err := x509.CreateCertificate(rand.Reader, &tmpl, rootCA, &priv.PublicKey, privKey)
+	if err != nil {
+		t.Fatal(err)
+	}
+	marshalCerts([][]byte{caBytes}, path.Join(dir, fmt.Sprintf("cert%s.pem", suffix)), t)
+	marshalKeys(priv, path.Join(dir, fmt.Sprintf("key%s.pem", suffix)), t)
+}
+
+func marshalCerts(caBytes [][]byte, certPath string, t *testing.T) {
+	var caPerm bytes.Buffer
+	for _, caBlock := range caBytes {
+		err := pem.Encode(&caPerm, &pem.Block{
+			Type:  "CERTIFICATE",
+			Bytes: caBlock,
+		})
+		if err != nil {
+			t.Fatal(err)
+		}
+	}
+	ioutil.WriteFile(certPath, caPerm.Bytes(), 0644)
+}
+
+func marshalKeys(privKey *ecdsa.PrivateKey, keyPath string, t *testing.T) {
+	privBytes, err := x509.MarshalECPrivateKey(privKey)
+	if err != nil {
+		t.Fatal(err)
+	}
+
+	var keyPerm bytes.Buffer
+	err = pem.Encode(&keyPerm, &pem.Block{
+		Type:  "EC PRIVATE KEY",
+		Bytes: privBytes,
+	})
+	if err != nil {
+		t.Fatal(err)
+	}
+	ioutil.WriteFile(keyPath, keyPerm.Bytes(), 0644)
+}
+
+func TestRootCAReload(t *testing.T) {
+	tmpdir, err := ioutil.TempDir(os.TempDir(), "tlsdir-reload")
+	if err != nil {
+		t.Fatal(err)
+	}
+	defer os.RemoveAll(tmpdir)
+	rootCAPath := path.Join(tmpdir, "ca-cert.pem")
+	logger := zap.NewExample()
+	rootCA, caBytes, privKey := createRootCertificateAuthority(rootCAPath, []byte{}, t)
+	generateCerts(privKey, rootCA, tmpdir, "_old", t)
+	tlsinfo := &TLSInfo{
+		TrustedCAFile:   rootCAPath,
+		CertFile:        path.Join(tmpdir, "cert_old.pem"),
+		KeyFile:         path.Join(tmpdir, "key_old.pem"),
+		ClientCertFile:  path.Join(tmpdir, "cert_old.pem"),
+		ClientKeyFile:   path.Join(tmpdir, "key_old.pem"),
+		Logger:          logger,
+		RefreshDuration: 100 * time.Millisecond,
+	}
+
+	ln, err := NewListener("127.0.0.1:0", "https", tlsinfo)
+	if err != nil {
+		t.Fatalf("unexpected NewListener error: %v", err)
+	}
+	defer ln.Close()
+	cfg, err := tlsinfo.ClientConfig()
+	if err != nil {
+		t.Fatal(err)
+	}
+	tr := &http.Transport{TLSClientConfig: cfg}
+	cli := &http.Client{Transport: tr}
+	go func() {
+		cli.Get("https://" + ln.Addr().String())
+	}()
+
+	conn, err := ln.Accept()
```

Comment:
Will this block infinitely if there is a bug? Could we avoid infinite blocking in tests, for example have some timeout?